tools/blobcp.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+    "flag"
+    "fmt"
+    "io"
+    "log"
+    "os"
+    "path/filepath"
+    "sync"
+    "sync/atomic"
+    "time"
+)
+
+var workers = flag.Int("n", 32, "Number of workers")
+var source = flag.String("s", "", "Source path to walk through")
+var dest = flag.String("d", "", "Destination to copy to")
+var include = flag.String("pattern", "", "Path pattern to match in source path")
+var processed int64 = 0
+
+func main() {
+    flag.Parse()
+    if *source == "" {
+        log.Printf("-s source path is missing")
+        flag.PrintDefaults()
+    } else if *dest == "" {
+        log.Printf("-d destination path is missing")
+        flag.PrintDefaults()
+    }
+
+    pending := sync.WaitGroup{}
+    files := make(chan string)
+
+    go func() {
+        last := processed
+        for {
+            <-time.After(5 * time.Second)
+            fmt.Printf("Processed %v, %v per second       \r", processed, (processed-last)/5)
+            last = processed
+        }
+    }()
+
+    start := time.Now()
+    pending.Add(1)
+    startWorkers(files, &pending)
+    submitFiles(*source, *include, files, &pending)
+    pending.Wait()
+
+    fmt.Printf("Processed %v files in %v sec", processed, time.Now().Sub(start))
+}
+
+func startWorkers(files chan string, pending *sync.WaitGroup) {
+    for i := 0; i < *workers; i++ {
+        go func() {
+            for p := range files {
+                processFile(p)
+                atomic.AddInt64(&processed, 1)
+                pending.Done()
+            }
+        }()
+    }
+    pending.Done()
```

Comment:
why startWorkers doing this? this should be done at the end of submitFiles instead